src/wrappers/msvc_wrapper.cpp

```diff
@@ -214,13 +214,19 @@ std::map<std::string, std::string> msvc_wrapper_t::get_relevant_env_vars() {
 }
 
 std::string msvc_wrapper_t::get_program_id() {
+  const auto& pathToCL = m_args[0];
+  const auto vsFolderPos = pathToCL.rfind("Microsoft Visual Studio");
+  if (vsFolderPos != std::string::npos) {
+    return HASH_VERSION + pathToCL.substr(vsFolderPos);
```

Comment:
Optionally, would it be possible to figure out how to prevent Visual Studio from capturing stderr? It seems strange to me that it can capture output from a subprocess of BuildCache (maybe I don't know Win32 coding well enough) - or is cl.exe somehow aware of that it's being invoked from Visual Studio and explicitly sends it's output to the IDE process?

Reply:
We chose the early exit for performance reasons: one fewer process start that would occur on every single cache lookup. Now, I don't yet have measurements on the effects of this change, I could grab some if you think it can help in the decision whether or not use the alternative you proposed.

Reply:
> [..] is cl.exe somehow aware that it's being invoked from Visual Studio and explicitly sends it's output to the IDE process?

Just did some extra reading on the topic, and appearently this exactly the case! Not much official info, but multiple forums [mention](https://stackoverflow.com/a/7390214) that clearing the enviroment variable `VS_UNICODE_OUTPUT` can turn it off. I'll give it try if it's still true with Visual Studio 2019.

Reply:
Yes, please investigate that option (unfortunately I don't have an MSVC setup readily available to test in ATM).

Reply:
On a related note: Should BuildCache use `VS_UNICODE_OUTPUT` for its output to produce correct results in Visual Studio? It seems to me that for a cache miss, BuildCache should disable `VS_UNICODE_OUTPUT` to capture all output into the cache entry, and for a cache hit, BuildCache should output the result to `VS_UNICODE_OUTPUT`. Right?

Reply:
I can confirm that clearing `VS_UNICODE_OUTPUT` works, so there might not be need looking at the path at all.

Still, the question of performance is worth investigating. 

I've built the static libs of our project, ~350 translation units in Debug x64 - first filled up the cache, then measured the total build time with 100% cache hits. On a desktop PC with 4C/8T, the build took 60 secs when spawning `cl.exe` processes, and only 56 secs when hashing based on the path exclusively. Also used the pref feature of BuildCache, it revealed that an extra process start takes ~50ms on average on my machine.

I'm eager to hear your take on this. On one hand now we can have the more robust response from `cl.exe` inside Visual Studio too, on the other hand the difference in latency is not negligible.

Reply:
I think that correctness is more important than speed (possibly modulo the `BUILDCACHE_ACCURACY` setting).

The dimension that is easily forgotten is the shared L2 cache: When using a shared cache between different clients, it's important to:
* Not falsely get cache hits for incompatible software installations.
* Not get cache misses due to different installation paths and such.

When a shared cache is used, you can get an order of magnitude performance increase in an environment with just a handful of clients building the same code base (e.g. an office with several developers working on the same project, or in a CI system). In my book, that outweighs a few percent of performance increase for the local case.

So, on principle I like the `VS_UNICODE_OUTPUT` solution better than the installation-path solution, since it seems to provide the more accurate compiler identification. If process-start performance is a concern, I think that it should be tackled as a separate task. For instance it may be feasible to cache the compiler signature for a few seconds during a build session (e.g. with the compiler path and/or include dirs etc as a cache key).

Reply:
Thanks for the insight, added the `VS_UNICODE_OUTPUT` solution in https://github.com/mbitsnbites/buildcache/pull/114/commits/c09e8a4e663f009ed24967ac2e57aa670db7dfd6. On your "related note": indeed, full correctness would involve disabling redirection to IDE on cache miss, and enabling on cache hit. For now, I chose to disable it in `get_program_id()` (which is obvious), and in the preprocessing step, which needs some explanation: 

Regulary, upon a successful compilation, `cl.exe` outputs the file name on stdout - which BuildCache correctly captures. With `/EP` flag though, stdout is occupied by the preprocessing output, and the file name appears on stderr - otherwise redirected to Visual Studio. So this fixes an annoyance that file names could appeare twice in the output window.